From ef4c96389c6b73587946e5c7ee69cacb6dafa8d0 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Maxime=20M=C3=A9r=C3=A9?= <maxime.mere@foss.st.com>
Date: Mon, 24 Feb 2025 13:21:48 +0100
Subject: [PATCH] hwrng: stm32 - fix RNG init for STM32MP21
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

RNG peripheral of STM32MP21 add the bit BUSY in the register RNG_SR.
Add support of this bit and fix error verification in stm32_rng_init().

Signed-off-by: Thomas Bourgoin <thomas.bourgoin@foss.st.com>
Signed-off-by: Maxime Méré <maxime.mere@foss.st.com>
Change-Id: Ibefb7b30270469964055562aaa481446a40d229f
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/429868
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Yann GAUTIER <yann.gautier@foss.st.com>
---
 drivers/char/hw_random/stm32-rng.c | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/drivers/char/hw_random/stm32-rng.c b/drivers/char/hw_random/stm32-rng.c
index 9c630623c53d..0cb084b35062 100644
--- a/drivers/char/hw_random/stm32-rng.c
+++ b/drivers/char/hw_random/stm32-rng.c
@@ -35,6 +35,7 @@
 #define RNG_SR_DRDY		BIT(0)
 #define RNG_SR_CECS		BIT(1)
 #define RNG_SR_SECS		BIT(2)
+#define RNG_SR_BUSY		BIT(4)
 #define RNG_SR_CEIS		BIT(5)
 #define RNG_SR_SEIS		BIT(6)
 #define RNG_SR_ERROR_MASK	(RNG_SR_CECS | RNG_SR_SECS | RNG_SR_CEIS | RNG_SR_SEIS)
@@ -284,7 +285,7 @@ static int stm32_rng_init(struct hwrng *rng)
 	struct stm32_rng_private *priv =
 	    container_of(rng, struct stm32_rng_private, rng);
 	int err;
-	u32 reg;
+	u32 reg, mask;
 
 	err = clk_prepare_enable(priv->clk);
 	if (err)
@@ -359,7 +360,10 @@ static int stm32_rng_init(struct hwrng *rng)
 	err = readl_relaxed_poll_timeout_atomic(priv->base + RNG_SR, reg,
 						reg & RNG_SR_DRDY,
 						10, 100000);
-	if (err || (reg & RNG_SR_ERROR_MASK)) {
+
+	mask = RNG_SR_ERROR_MASK;
+
+	if (err || (reg & mask)) {
 		if (priv->bus_clk)
 			clk_disable_unprepare(priv->bus_clk);
 		clk_disable_unprepare(priv->clk);
-- 
2.50.1

